src/scraper.rs

```diff
@@ -0,0 +1,392 @@
+            "Right" => Self::Right,
+            _ => bail!("unexpected political affiliation shorthand: {}", s),
+        };
+
+        Ok(x)
+    }
+}
+
+impl Side {
+    pub fn emoji(&self) -> &'static str {
+        match *self {
+            Side::Left => "🟦",
+            Side::CenterLeft => "🔵",
+            Side::Center => "🟣",
+            Side::CenterRight => "🔴",
+            Side::Right => "🟥",
+        }
+    }
+}
+
+impl Paragraph<'_> {
+    pub fn text(&self) -> String {
+        self.0.text()
+    }
+
+    pub fn telegram_html(&self) -> String {
+        self.0
+            .children()
+            .map(|n| {
+                Some(n)
+                    .filter(|n| n.is(Name("a")))
+                    .and_then(|a| {
+                        Some(format!(
+                            r#"<a href="{url}">{text}</a>"#,
+                            url = a.attr("href")?,
+                            text = a.text()
+                        ))
+                    })
+                    .unwrap_or_else(|| n.text())
+            })
+            .collect::<String>()
+    }
+}
+
+pub trait FromHTML<'a>: 'a + Sized {
+    fn from_html(html: &'a Document) -> anyhow::Result<Self>;
+}
+
+impl FromHTML<'_> for MainPage {
+    // TODO: get list of stories directly from AllSides API instead of this atrocity
+    fn from_html(html: &Document) -> anyhow::Result<Self> {
+        let mut teasers: Vec<_> = html.find(Class("view-story-id-single-story")).map(|block| {
+            let url = block.find(Name("a")).next()
+                .and_then(|node| node.attr("href"))
+                .ok_or_else(|| anyhow!("cannot query story url (class=view-story-id-single-story > a.href)"))?;
+
+            let title = block.find(Class("story-title")).next()
+                .ok_or_else(|| anyhow!("cannot query story url (class=view-story-id-single-story > class=story-title)"))?
+                .text();
+
+            let img_url = block.find(Class("story-id-image")).next()
+                .and_then(|node| node.find(Name("img")).next())
+                .and_then(|node| node.attr("src"))
+                .ok_or_else(|| anyhow!("cannot query story url (class=view-story-id-single-story > 0 > img.src)"))?;
+
+            Ok(Teaser {
+                title,
+                url: normalize_allsides_url(url),
+                img_url: img_url.to_owned(),
+            })
+        }).collect::<Result<_, anyhow::Error>>()?;
+
+        // Reverse posts for them to be in chronological order
+        teasers.reverse();
+
+        if teasers.is_empty() {
+            bail!("the main page contains no stories: parsing is broken");
+        }
+
+        Ok(MainPage { teasers })
+    }
+}
+
+impl<'a> FromHTML<'a> for Story<'a> {
+    fn from_html(html: &'a Document) -> anyhow::Result<Self> {
+        let story = html
+            .find(Attr("id", "content"))
+            .next()
+            .ok_or_else(|| anyhow!("cannot query story body (id=content)"))?;
```

Comment:
Given how often `.ok_or_else(|| anyhow!(<string literal>))` occurs, does make sense to make an extension trait for `Result` to provide a shortcut?

Reply:
Since those aren't actually using formatting, yeah, good idea